File: airbyte-server/src/main/java/io/airbyte/server/apis/DestinationDefinitionApiController.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis;

import static io.airbyte.commons.auth.AuthRoleConstants.ADMIN;
import static io.airbyte.commons.auth.AuthRoleConstants.AUTHENTICATED_USER;
import static io.airbyte.commons.auth.AuthRoleConstants.EDITOR;
import static io.airbyte.commons.auth.AuthRoleConstants.ORGANIZATION_EDITOR;
import static io.airbyte.commons.auth.AuthRoleConstants.ORGANIZATION_READER;
import static io.airbyte.commons.auth.AuthRoleConstants.READER;
import static io.airbyte.commons.auth.AuthRoleConstants.WORKSPACE_EDITOR;
import static io.airbyte.commons.auth.AuthRoleConstants.WORKSPACE_READER;

import io.airbyte.api.generated.DestinationDefinitionApi;
import io.airbyte.api.model.generated.ActorDefinitionIdWithScope;
import io.airbyte.api.model.generated.CustomDestinationDefinitionCreate;
import io.airbyte.api.model.generated.DestinationDefinitionIdRequestBody;
import io.airbyte.api.model.generated.DestinationDefinitionIdWithWorkspaceId;
import io.airbyte.api.model.generated.DestinationDefinitionRead;
import io.airbyte.api.model.generated.DestinationDefinitionReadList;
import io.airbyte.api.model.generated.DestinationDefinitionUpdate;
import io.airbyte.api.model.generated.PrivateDestinationDefinitionRead;
import io.airbyte.api.model.generated.PrivateDestinationDefinitionReadList;
import io.airbyte.api.model.generated.ScopeType;
import io.airbyte.api.model.generated.WorkspaceIdRequestBody;
import io.airbyte.commons.auth.SecuredWorkspace;
import io.airbyte.commons.server.handlers.DestinationDefinitionsHandler;
import io.airbyte.commons.server.scheduling.AirbyteTaskExecutors;
import io.airbyte.commons.server.validation.ActorDefinitionAccessValidator;
import io.micronaut.context.annotation.Context;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.annotation.Status;
import io.micronaut.scheduling.annotation.ExecuteOn;
import io.micronaut.security.annotation.Secured;
import io.micronaut.security.rules.SecurityRule;

@Controller("/api/v1/destination_definitions")
@Context
@Secured(SecurityRule.IS_AUTHENTICATED)
public class DestinationDefinitionApiController implements DestinationDefinitionApi {

  private final DestinationDefinitionsHandler destinationDefinitionsHandler;
  private final ActorDefinitionAccessValidator accessValidator;

  public DestinationDefinitionApiController(final DestinationDefinitionsHandler destinationDefinitionsHandler,
                                            final ActorDefinitionAccessValidator accessValidator) {
    this.destinationDefinitionsHandler = destinationDefinitionsHandler;
    this.accessValidator = accessValidator;
  }

  @Post(uri = "/create_custom")
  @Secured({EDITOR, WORKSPACE_EDITOR, ORGANIZATION_EDITOR})
  @SecuredWorkspace
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionRead createCustomDestinationDefinition(final CustomDestinationDefinitionCreate customDestinationDefinitionCreate) {
    // legacy calls contain workspace id instead of scope id and scope type
    if (customDestinationDefinitionCreate.getWorkspaceId() != null) {
      customDestinationDefinitionCreate.setScopeType(ScopeType.WORKSPACE);
      customDestinationDefinitionCreate.setScopeId(customDestinationDefinitionCreate.getWorkspaceId());
    }
    return ApiHelper.execute(() -> destinationDefinitionsHandler.createCustomDestinationDefinition(customDestinationDefinitionCreate));
  }

  @Post(uri = "/delete")
  // the accessValidator will provide additional authorization checks, depending on Airbyte edition.
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  @Status(HttpStatus.NO_CONTENT)
  public void deleteDestinationDefinition(final DestinationDefinitionIdRequestBody destinationDefinitionIdRequestBody) {
    accessValidator.validateWriteAccess(destinationDefinitionIdRequestBody.getDestinationDefinitionId());
    ApiHelper.execute(() -> {
      destinationDefinitionsHandler.deleteDestinationDefinition(destinationDefinitionIdRequestBody);
      return null;
    });
  }

  @Post(uri = "/get")
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionRead getDestinationDefinition(final DestinationDefinitionIdRequestBody destinationDefinitionIdRequestBody) {
    return ApiHelper.execute(() -> destinationDefinitionsHandler.getDestinationDefinition(destinationDefinitionIdRequestBody));
  }

  @Post("/get_for_scope")
  @Secured({READER, WORKSPACE_READER, ORGANIZATION_READER})
  @SecuredWorkspace
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionRead getDestinationDefinitionForScope(final ActorDefinitionIdWithScope actorDefinitionIdWithScope) {
    return ApiHelper.execute(() -> destinationDefinitionsHandler.getDestinationDefinitionForScope(actorDefinitionIdWithScope));
  }

  @SuppressWarnings("LineLength")
  @Post(uri = "/get_for_workspace")
  @Secured({READER, WORKSPACE_READER, ORGANIZATION_READER})
  @SecuredWorkspace
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionRead getDestinationDefinitionForWorkspace(
                                                                        final DestinationDefinitionIdWithWorkspaceId destinationDefinitionIdWithWorkspaceId) {
    return ApiHelper.execute(() -> destinationDefinitionsHandler.getDestinationDefinitionForWorkspace(destinationDefinitionIdWithWorkspaceId));
  }

  @SuppressWarnings("LineLength")
  @Post(uri = "/grant_definition")
  @Secured({ADMIN})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public PrivateDestinationDefinitionRead grantDestinationDefinition(final ActorDefinitionIdWithScope actorDefinitionIdWithScope) {
    return ApiHelper.execute(() -> destinationDefinitionsHandler.grantDestinationDefinitionToWorkspaceOrOrganization(actorDefinitionIdWithScope));
  }

  @Post(uri = "/list")
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionReadList listDestinationDefinitions() {
    return ApiHelper.execute(destinationDefinitionsHandler::listDestinationDefinitions);
  }

  @Post(uri = "/list_for_workspace")
  @Secured({READER, WORKSPACE_READER, ORGANIZATION_READER})
  @SecuredWorkspace
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionReadList listDestinationDefinitionsForWorkspace(final WorkspaceIdRequestBody workspaceIdRequestBody) {
    return ApiHelper.execute(() -> destinationDefinitionsHandler.listDestinationDefinitionsForWorkspace(workspaceIdRequestBody));
  }

  @Post(uri = "/list_latest")
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionReadList listLatestDestinationDefinitions() {
    return ApiHelper.execute(destinationDefinitionsHandler::listLatestDestinationDefinitions);
  }

  @Post(uri = "/list_private")
  @Secured({ADMIN})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public PrivateDestinationDefinitionReadList listPrivateDestinationDefinitions(final WorkspaceIdRequestBody workspaceIdRequestBody) {
    return ApiHelper.execute(() -> destinationDefinitionsHandler.listPrivateDestinationDefinitions(workspaceIdRequestBody));
  }

  @Post(uri = "/revoke_definition")
  @Secured({ADMIN})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public void revokeDestinationDefinition(final ActorDefinitionIdWithScope actorDefinitionIdWithScope) {
    ApiHelper.execute(() -> {
      destinationDefinitionsHandler.revokeDestinationDefinition(actorDefinitionIdWithScope);
      return null;
    });
  }

  @Post(uri = "/update")
  // the accessValidator will provide additional authorization checks, depending on Airbyte edition.
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public DestinationDefinitionRead updateDestinationDefinition(final DestinationDefinitionUpdate destinationDefinitionUpdate) {
    accessValidator.validateWriteAccess(destinationDefinitionUpdate.getDestinationDefinitionId());
    return ApiHelper.execute(() -> destinationDefinitionsHandler.updateDestinationDefinition(destinationDefinitionUpdate));
  }

}

File: airbyte-server/src/main/kotlin/io/airbyte/server/apis/controllers/DestinationDefinitionApiController.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis.controllers

import io.airbyte.api.generated.DestinationDefinitionApi
import io.airbyte.api.model.generated.ActorDefinitionIdWithScope
import io.airbyte.api.model.generated.CustomDestinationDefinitionCreate
import io.airbyte.api.model.generated.DestinationDefinitionIdRequestBody
import io.airbyte.api.model.generated.DestinationDefinitionIdWithWorkspaceId
import io.airbyte.api.model.generated.DestinationDefinitionRead
import io.airbyte.api.model.generated.DestinationDefinitionReadList
import io.airbyte.api.model.generated.DestinationDefinitionUpdate
import io.airbyte.api.model.generated.PrivateDestinationDefinitionRead
import io.airbyte.api.model.generated.PrivateDestinationDefinitionReadList
import io.airbyte.api.model.generated.ScopeType
import io.airbyte.api.model.generated.WorkspaceIdRequestBody
import io.airbyte.commons.auth.AuthRoleConstants
import io.airbyte.commons.auth.generated.Intent
import io.airbyte.commons.auth.permissions.RequiresIntent
import io.airbyte.commons.server.handlers.DestinationDefinitionsHandler
import io.airbyte.commons.server.scheduling.AirbyteTaskExecutors
import io.airbyte.commons.server.validation.ActorDefinitionAccessValidator
import io.airbyte.server.apis.execute
import io.micronaut.context.annotation.Context
import io.micronaut.http.HttpStatus
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Post
import io.micronaut.http.annotation.Status
import io.micronaut.scheduling.annotation.ExecuteOn
import io.micronaut.security.annotation.Secured
import io.micronaut.security.rules.SecurityRule
import java.util.concurrent.Callable

@Controller("/api/v1/destination_definitions")
@Context
@Secured(SecurityRule.IS_AUTHENTICATED)
open class DestinationDefinitionApiController(
  private val destinationDefinitionsHandler: DestinationDefinitionsHandler,
  private val accessValidator: ActorDefinitionAccessValidator,
) : DestinationDefinitionApi {
  @Post(uri = "/create_custom")
  @RequiresIntent(Intent.UploadCustomConnector)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun createCustomDestinationDefinition(
    @Body customDestinationDefinitionCreate: CustomDestinationDefinitionCreate,
  ): DestinationDefinitionRead? {
    // legacy calls contain workspace id instead of scope id and scope type
    if (customDestinationDefinitionCreate.workspaceId != null) {
      customDestinationDefinitionCreate.setScopeType(ScopeType.WORKSPACE)
      customDestinationDefinitionCreate.scopeId = customDestinationDefinitionCreate.workspaceId
    }
    return execute {
      destinationDefinitionsHandler.createCustomDestinationDefinition(
        customDestinationDefinitionCreate,
      )
    }
  }

  @Post(uri = "/delete") // the accessValidator will provide additional authorization checks, depending on Airbyte edition.
  @Secured(AuthRoleConstants.AUTHENTICATED_USER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Status(HttpStatus.NO_CONTENT)
  override fun deleteDestinationDefinition(
    @Body destinationDefinitionIdRequestBody: DestinationDefinitionIdRequestBody,
  ) {
    accessValidator.validateWriteAccess(destinationDefinitionIdRequestBody.destinationDefinitionId)
    execute<Any?> {
      destinationDefinitionsHandler.deleteDestinationDefinition(destinationDefinitionIdRequestBody.destinationDefinitionId)
      null
    }
  }

  @Post(uri = "/get")
  @Secured(AuthRoleConstants.AUTHENTICATED_USER, AuthRoleConstants.DATAPLANE)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun getDestinationDefinition(
    @Body destinationDefinitionIdRequestBody: DestinationDefinitionIdRequestBody,
  ): DestinationDefinitionRead? =
    execute {
      destinationDefinitionsHandler.getDestinationDefinition(
        destinationDefinitionIdRequestBody.destinationDefinitionId,
        true,
      )
    }

  @Post("/get_for_scope")
  @Secured(AuthRoleConstants.WORKSPACE_READER, AuthRoleConstants.ORGANIZATION_READER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun getDestinationDefinitionForScope(
    @Body actorDefinitionIdWithScope: ActorDefinitionIdWithScope,
  ): DestinationDefinitionRead? =
    execute {
      destinationDefinitionsHandler.getDestinationDefinitionForScope(
        actorDefinitionIdWithScope,
      )
    }

  @Post(uri = "/get_for_workspace")
  @Secured(AuthRoleConstants.WORKSPACE_READER, AuthRoleConstants.ORGANIZATION_READER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun getDestinationDefinitionForWorkspace(
    @Body destinationDefinitionIdWithWorkspaceId: DestinationDefinitionIdWithWorkspaceId,
  ): DestinationDefinitionRead? =
    execute {
      destinationDefinitionsHandler.getDestinationDefinitionForWorkspace(
        destinationDefinitionIdWithWorkspaceId,
      )
    }

  @Post(uri = "/grant_definition")
  @Secured(AuthRoleConstants.ADMIN)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun grantDestinationDefinition(
    @Body actorDefinitionIdWithScope: ActorDefinitionIdWithScope,
  ): PrivateDestinationDefinitionRead? =
    execute {
      destinationDefinitionsHandler.grantDestinationDefinitionToWorkspaceOrOrganization(
        actorDefinitionIdWithScope,
      )
    }

  @Post(uri = "/list")
  @Secured(AuthRoleConstants.AUTHENTICATED_USER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun listDestinationDefinitions(): DestinationDefinitionReadList? =
    execute(
      Callable {
        destinationDefinitionsHandler.listDestinationDefinitions()
      },
    )

  @Post(uri = "/list_for_workspace")
  @Secured(AuthRoleConstants.WORKSPACE_READER, AuthRoleConstants.ORGANIZATION_READER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun listDestinationDefinitionsForWorkspace(
    @Body workspaceIdRequestBody: WorkspaceIdRequestBody,
  ): io.airbyte.api.model.generated.DestinationDefinitionReadList? =
    execute {
      destinationDefinitionsHandler.listDestinationDefinitionsForWorkspace(
        workspaceIdRequestBody,
      )
    }

  @Post(uri = "/list_latest")
  @Secured(AuthRoleConstants.AUTHENTICATED_USER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun listLatestDestinationDefinitions(): DestinationDefinitionReadList? =
    execute(
      Callable {
        destinationDefinitionsHandler.listLatestDestinationDefinitions()
      },
    )

  @Post(uri = "/list_private")
  @Secured(AuthRoleConstants.ADMIN)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun listPrivateDestinationDefinitions(
    @Body workspaceIdRequestBody: WorkspaceIdRequestBody,
  ): PrivateDestinationDefinitionReadList? =
    execute {
      destinationDefinitionsHandler.listPrivateDestinationDefinitions(
        workspaceIdRequestBody,
      )
    }

  @Post(uri = "/revoke_definition")
  @Secured(AuthRoleConstants.ADMIN)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun revokeDestinationDefinition(
    @Body actorDefinitionIdWithScope: ActorDefinitionIdWithScope,
  ) {
    execute<Any?> {
      destinationDefinitionsHandler.revokeDestinationDefinition(actorDefinitionIdWithScope)
      null
    }
  }

  @Post(uri = "/update") // the accessValidator will provide additional authorization checks, depending on Airbyte edition.
  @Secured(AuthRoleConstants.AUTHENTICATED_USER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun updateDestinationDefinition(
    @Body destinationDefinitionUpdate: DestinationDefinitionUpdate,
  ): DestinationDefinitionRead? {
    accessValidator.validateWriteAccess(destinationDefinitionUpdate.destinationDefinitionId)
    return execute {
      destinationDefinitionsHandler.updateDestinationDefinition(
        destinationDefinitionUpdate,
      )
    }
  }
}
